kge/job/train.py

```diff
@@ -963,22 +991,50 @@ def _process_batch(self, batch_index, batch) -> TrainingJob._ProcessBatchResult:
                     forward_time += time.time()
 
-                    # determine indexes of relevant scores in scoring matrix
-                    prepare_time -= time.time()
-                    row_indexes = (
-                        torch.arange(chunk_size, device=self.device)
-                        .unsqueeze(1)
-                        .repeat(1, 1 + num_samples)
-                        .view(-1)
-                    )  # 000 111 222; each 1+num_negative times (here: 3)
-                    prepare_time += time.time()
+                    if not self._shared_raw_results:
+                        # determine indexes of relevant scores in scoring matrix
+                        prepare_time -= time.time()
+                        row_indexes = (
+                            torch.arange(chunk_size, device=self.device)
+                            .unsqueeze(1)
+                            .repeat(1, 1 + num_samples)
+                            .view(-1)
+                        )  # 000 111 222; each 1+num_negative times (here: 3)
+                        prepare_time += time.time()
+
+                        # now pick the scores we need
+                        forward_time -= time.time()
+                        scores = all_scores[row_indexes, column_indexes].view(
+                            chunk_size, -1
+                        )
+                        forward_time += time.time()
+                    else:
+                        # create the complete scoring matrix
+                        scores = torch.empty(
+                            chunk_size, num_samples + 1, device=self.device
+                        )
 
-                    # now pick the scores we need
-                    forward_time -= time.time()
-                    scores = all_scores[row_indexes, column_indexes].view(
-                        chunk_size, -1
-                    )
-                    forward_time += time.time()
+                        # fill in the unique negative scores. first column is left empty
+                        # to hold positive scores
+                        num_unique = len(unique_targets) - 1
+                        scores[:, 1 : (num_unique + 1)] = all_scores[:, :-1]
+                        drop_rows = torch.nonzero(
+                            drop_index[slot] != num_unique, as_tuple=False
+                        ).squeeze()
+                        scores[drop_rows, drop_index[slot][drop_rows] + 1] = all_scores[
+                            drop_rows, -1
+                        ]
+
+                        # repeat scores as needed for WR sampling
+                        if num_unique != num_samples:
+                            scores[:, (num_unique + 1) :] = all_scores[
+                                :, repeat_indexes[slot]
+                            ]
+
+                        # fill in positive scores
+                        scores[:, 0] = self.model.score_spo(
```

Comment:
Use S, P, O instead of 0,1,2

We could also compute this once before looping over the slots. The only problem with that is, that we have to retain the graph then.

Reply:
This 0 refers to the positive column, not to S. Compute once is a good idea to try!

Reply:
Seems like my review ended up a line above. I meant triples[:, 0], triples[: ,1], triples[:, 2] the line below